GVFS/FastFetch/NativeUnixMethods.cs

```diff
@@ -14,21 +14,35 @@ public class NativeUnixMethods
         public const int Create = 0x0200;
         public const int Truncate = 0x0400;
         private const int InvalidFileDescriptor = -1;
+        private const ushort SymLinkMode = 0xA000;
 
         public static unsafe void WriteFile(ITracer tracer, byte* originalData, long originalSize, string destination, ushort mode)
         {
             int fileDescriptor = InvalidFileDescriptor;
             try
             {
-                fileDescriptor = Open(destination, WriteOnly | Create | Truncate, mode);
-                if (fileDescriptor != InvalidFileDescriptor)
+                if (mode == SymLinkMode)
                 {
-                    IntPtr result = Write(fileDescriptor, originalData, (IntPtr)originalSize);
-                    if (result.ToInt32() == -1)
+                    string linkLocation = Marshal.PtrToStringUTF8(new IntPtr(originalData));
+                    int result = CreateSymLink(linkLocation, destination);
+                    if (result == -1)
                     {
+                        tracer.RelatedInfo("Failed to create symlink({0},{1})", linkLocation, destination);
                         throw new Win32Exception(Marshal.GetLastWin32Error());
                     }
                 }
+                else
+                {
+                    fileDescriptor = Open(destination, WriteOnly | Create | Truncate, mode);
+                    if (fileDescriptor != InvalidFileDescriptor)
+                    {
+                        IntPtr result = Write(fileDescriptor, originalData, (IntPtr)originalSize);
+                        if (result.ToInt32() == -1)
+                        {
+                            throw new Win32Exception(Marshal.GetLastWin32Error());

```

Comment:
Include a message in this `Win32Exception ` that it was the `Write` that failed?